hoomd/hpmc/shape_move.py

```diff
@@ -65,215 +72,152 @@ class Constant(ShapeMove):
+            the shape definition of the integrator:
 
-        callback (Callback): The python class that will be called
-            to update the particle shapes
+                ``fun(typeid, param_list) -> dict``
 
-        params (dict): Dictionary of types and the corresponding list
-            of initial parameters to pass to the callback
-            (ex: {'A' : [1.0], 'B': [0.0]})
+        params (`TypeParameter` [``particle type``, `list`]): List of tunable
+            parameters to be updated.
 
-        param_move_probability (float): Average fraction of parameters to change during
-            each shape move
+        move_probability (`float`): Average fraction of shape parameters to
+            change each timestep.
     """
 
-    def __init__(self, callback, params, param_move_probability):
-        # TODO: params should be implemented as TypeParameter
-        param_dict = ParameterDict(
-            callback=Callback,
-            params=dict(params),
-            param_move_probability=float(param_move_probability))
-        param_dict["callback"] = callback
-        self._param_dict.update(param_dict)
+    _suported_shapes = {
+        'ConvexPolyhedron', 'ConvexSpheropolyhedron', 'Ellipsoid'
+    }
 
-    def _attach(self):
-        integrator = self._simulation.operations.integrator
-        if not isinstance(integrator, integrate.HPMCIntegrator):
-            raise RuntimeError("The integrator must be a HPMC integrator.")
-        if not integrator._attached:
-            raise RuntimeError("Integrator is not attached yet.")
+    def __init__(self, move_probability=1):
 
-        move_cls = None
-        shapes = [
-            'Sphere', 'ConvexPolygon', 'SimplePolygon', 'ConvexPolyhedron',
-            'ConvexSpheropolyhedron', 'Ellipsoid', 'ConvexSpheropolygon',
-            'Polyhedron', 'Sphinx', 'SphereUnion'
-        ]
-        print()
-        for shape in shapes:
-            if isinstance(integrator, getattr(integrate, shape)):
-                move_cls = getattr(_hpmc, 'PythonShapeMove' + shape)
-        if move_cls is None:
-            raise RuntimeError("Integrator not supported")
+        super().__init__(move_probability)
+        param_dict = ParameterDict(callback=object)
+        self._param_dict.update(param_dict)
 
-        ntypes = self._simulation.state._cpp_sys_def.getParticleData(
-        ).getNTypes()
-        self._cpp_obj = move_cls(self._simulation.state._cpp_sys_def, ntypes,
-                                 self.callback, self.params,
-                                 self.param_move_probability)
-        super()._attach()
+        typeparam_shapeparams = TypeParameter('params',
+                                              type_kind='particle_types',
+                                              param_dict=TypeParameterDict(
+                                                [float], len_keys=1))
+        self._add_typeparam(typeparam_shapeparams)
```

Comment:
How would that be done? I'm not sure I follow.

Reply:
Like
``` python
class PythonMove:
    def __init__(self):
        self.type_params = {"A": {"foo": ...}, "B": {"foo": ...}}

    def __call__(self, type):
        ...
```
Users can store anything they want in a class.

Reply:
I see what you mean now but this would only work if the trial move was completely performed in python, right? (https://github.com/glotzerlab/hoomd-blue/pull/989#discussion_r848797420)